File: golang/examples/consumer.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"qbus"
)

// GoCallback 消费者的回调，可携带用户自定义数据，比如这里的 Consumer
type GoCallback struct {
	// 仅用于 DeliveryMsgForCommitOffset 才需要这个字段
	Consumer qbus.QbusConsumer
}

// DeliveryMsg 默认的回调函数
func (cb *GoCallback) DeliveryMsg(topic string, msg string, msgLen int64) {
	fmt.Printf("Topic: %s | msg: %s\n", topic, msg)
}

// DeliveryMsgForCommitOffset 用户自行管理并提交 offset 或确认消息的回调函数，需要修改默认配置
func (cb *GoCallback) DeliveryMsgForCommitOffset(info qbus.QbusMsgContentInfo) {
	fmt.Printf("User commit offset | Topic: %s | msg: %s\n", info.GetTopic(), info.GetMsg())
	cb.Consumer.CommitOffset(info)
}

func main() {
	if len(os.Args) < 5 {
		fmt.Printf("Usage: ./consumer config_path topic_name group_name cluster_name\n")
		return
	}

	configPath := os.Args[1]
	topicName := os.Args[2]
	group := os.Args[3]
	clusterName := os.Args[4]

	fmt.Printf("topic: %s | group: %s | cluster: %s\n", topicName, group, clusterName)

	sigs := make(chan os.Signal, 1)
	done := make(chan bool, 1)

	signal.Notify(sigs, os.Interrupt)

	go func() {
		sig := <-sigs
		fmt.Println()
		fmt.Println(sig)
		done <- true
	}()

	var goCallback GoCallback
	callback := qbus.NewDirectorQbusConsumerCallback(&goCallback)
	defer qbus.DeleteQbusConsumerCallback(callback)

	consumer := qbus.NewQbusConsumer()
	defer qbus.DeleteQbusConsumer(consumer)

	goCallback.Consumer = consumer

	if !consumer.Init(clusterName, "consumer.log", configPath, callback) {
		fmt.Println("Init failed")
		os.Exit(1)
	}

	if !consumer.SubscribeOne(group, topicName) {
		fmt.Println("SubscribeOne failed")
		os.Exit(2)
	}

	if !consumer.Start() {
		fmt.Println("Start failed")
		os.Exit(3)
	}

	<-done
	consumer.Stop()
	fmt.Println("done")
}

File: python/examples/consumer.py
import qbus
import signal
import sys

class Callback(qbus.QbusConsumerCallback):
    def __init__(self):
        qbus.QbusConsumerCallback.__init__(self)

    def setConsumer(self, consumer):
        self.consumer = consumer

    def deliveryMsg(self, topic, msg, msg_len):
        print "Topic: %s | msg: %s" % (topic, msg)

    def deliveryMsgForCommitOffset(self, info):
        print "User commit offset | Topic: %s | msg: %s" % (info.topic, info.msg)
        self.consumer.CommitOffset(info)


if len(sys.argv) < 5:
    print "Usage: ./consumer config_path topic_name group_name cluster_name"
    sys.exit(1)

config_path = sys.argv[1]
topic_name = sys.argv[2]
group = sys.argv[3]
cluster_name = sys.argv[4]

print "topic: %s | group: %s | cluster: %s" % (topic_name, group, cluster_name)

callback = Callback().__disown__()
consumer = qbus.QbusConsumer()
callback.setConsumer(consumer)

if False == consumer.init(cluster_name, "consumer.log", config_path, callback):
    print "Init failed"
    sys.exit(2)

if False == consumer.subscribeOne(group, topic_name):
    print "SubscribeOne failed"
    sys.exit(3)

if False == consumer.start():
    print "Start failed"
    sys.exit(4)

try:
    signal.pause()
except KeyboardInterrupt:
    consumer.stop()
    print "done"
